tree/ntupleutil/v7/src/RNTupleExporter.cxx

```diff
@@ -0,0 +1,124 @@
+#include <ROOT/RClusterPool.hxx>
+#include <ROOT/RLogger.hxx>
+#include <fstream>
+#include <sstream>
+
+namespace ROOT::Experimental::Internal {
+
+struct RColumnExportInfo {
+   const RColumnDescriptor *fColDesc;
+   const RFieldDescriptor *fFieldDesc;
+   std::string fQualifiedName;
+
+   RColumnExportInfo(const RNTupleDescriptor &desc, const RColumnDescriptor &colDesc, const RFieldDescriptor &fieldDesc)
+      : fColDesc(&colDesc),
+        fFieldDesc(&fieldDesc),
+        fQualifiedName(desc.GetQualifiedFieldName(fieldDesc.GetId()) + '-' + std::to_string(colDesc.GetIndex()))
+   {
+   }
+};
+
+static void AddColumnsFromField(std::vector<RColumnExportInfo> &vec, const RNTupleDescriptor &desc,
+                                const RFieldDescriptor &fieldDesc)
+{
+   R__LOG_DEBUG(1) << "processing field \"" << desc.GetQualifiedFieldName(fieldDesc.GetId()) << "\"";
+
+   for (const auto &subfieldDesc : desc.GetFieldIterable(fieldDesc)) {
+      if (subfieldDesc.IsProjectedField())
+         continue;
+
+      for (const auto &colDesc : desc.GetColumnIterable(subfieldDesc)) {
+         vec.emplace_back(desc, colDesc, subfieldDesc);
+      }
+      AddColumnsFromField(vec, desc, subfieldDesc);
+   }
+}
+
+RExportPagesResult RNTupleExporter::ExportPages(RPageSource &source, const RExportPagesOptions &options)
+{
+   RExportPagesResult res = {};
+
+   // make sure the source is attached
+   source.Attach();
+
+   auto desc = source.GetSharedDescriptorGuard();
+   RClusterPool clusterPool{source};
+
+   // Collect column info
+   std::vector<RColumnExportInfo> columnInfos;
+   AddColumnsFromField(columnInfos, desc.GetRef(), desc->GetFieldZero());
+
+   // Collect ColumnSet for the cluster pool query
+   RCluster::ColumnSet_t columnSet;
+   columnSet.reserve(columnInfos.size());
+   for (const auto &colInfo : columnInfos) {
+      columnSet.emplace(colInfo.fColDesc->GetPhysicalId());
+   }
+
+   bool reportFilenames = (options.fFlags & RExportPagesOptions::kReportExportedFileNames) != 0;
+
+   // Iterate over the clusters in order and dump pages
+   DescriptorId_t clusterId = desc->FindClusterId(0, 0);
+   while (clusterId != kInvalidDescriptorId) {
+      const auto &clusterDesc = desc->GetClusterDescriptor(clusterId);
+      const RCluster *cluster = clusterPool.GetCluster(clusterId, columnSet);
+      for (const auto &colInfo : columnInfos) {
+         DescriptorId_t columnId = colInfo.fColDesc->GetPhysicalId();
+         const auto &pages = clusterDesc.GetPageRange(columnId);
+         const auto &colRange =clusterDesc.GetColumnRange(columnId);
+         std::uint64_t pageIdx = 0;
+         if (reportFilenames)
+            res.fExportedFileNames.reserve(res.fExportedFileNames.size() + pages.fPageInfos.size());
+
+         R__LOG_DEBUG(0) << "exporting column \"" << colInfo.fQualifiedName << "\" (" << pages.fPageInfos.size()
+                         << " pages)";
+
+         for (const auto &pageInfo : pages.fPageInfos) {
```

Comment:
I believe this works automatically, but maybe add an explicit check for `colRange.fIsSuppressed` in which case `pages.fPageInfos` should be `empty()`...